src/validationinterface.h

```diff
@@ -53,46 +58,110 @@ void CallFunctionInValidationInterfaceQueue(std::function<void ()> func);
  */
 void SyncWithValidationInterfaceQueue();
 
-class CValidationInterface {
+/**
+ * An interface to get callbacks about transactions entering and leaving
+ * mempool.
+ *
+ * Any class which extends both MempoolInterface and CValidationInterface will
+ * see all callbacks across both well-ordered (see individual callback text for
+ * details on the order guarantees).
+ *
+ * Callbacks called on a background thread have a separate order from those
+ * called on the thread generating the callbacks.
+ */
+class MempoolInterface {
 protected:
     /**
      * Protected destructor so that instances can only be deleted by derived classes.
      * If that restriction is no longer desired, this should be made public and virtual.
      */
-    ~CValidationInterface() = default;
+    ~MempoolInterface() = default;
     /**
-     * Notifies listeners of updated block chain tip
+     * Notifies listeners of a transaction having been added to mempool.
      *
      * Called on a background thread.
      */
-    virtual void UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownload) {}
+    virtual void TransactionAddedToMempool(const CTransactionRef &ptxn, const std::vector<CTransactionRef>& txn_replaced) {}
     /**
-     * Notifies listeners of a transaction having been added to mempool.
+     * Notifies listeners of a transaction leaving mempool.
+     *
+     * This only fires for transactions which leave mempool because of expiry,
+     * size limiting, or reorg (changes in lock times/coinbase maturity). This
+     * does not include any transactions which are included in
+     * MempoolUpdatedForBlockConnect or TransactionAddedToMempool(txn_replaced)
+     *
+     * reason == REORG is not ordered with BlockConnected/BlockDisconnected!
+     *
+     * Note that in some rare cases (eg mempool limiting) a
+     * TransactionRemovedFromMempool event may fire with no corresponding
+     * TransactionAddedToMempool event for the same transaction.
+     * (TODO: remove this edge case)
```

Comment:
Thanks for adding the documentation.
I think it'd be better to open an issue to track the TODOs, instead of putting them in the source code comments (this tends to get out of date very quickly).